### PR TITLE
fix(metrics): compute AUPR as step-wise average precision

### LINKED ISSUE #3781: AUPR ranks constant and all-zeros detectors above a random one, and the docstring example is wrong
`AUPR.compute()` integrates the precision-recall curve with `auc(rec, prec, reorder=True)`. A PR curve is not a function of recall, so sorting the points by recall and trapezoid-integrating does not give the area under it. The effect is large, and it is already visible in the module's own documentation.

## The module's own example is wrong

From the docstring at the top of `src/anomalib/metrics/aupr.py`:

```
>>> aupr = AUPR(fields=["pred_score", "gt_label"])
>>> aupr(batch)
tensor(0.8750)
```

with `pred_score=[0.1, 0.2, 0.8, 0.9]` and `gt_label=[0, 0, 1, 1]`. Every anomaly is ranked above every normal sample, so that is a perfect detector and the average precision is 1.0. The documented 0.8750 reproduces exactly.

## Measured

n=2000, prevalence 0.05, so the chance level is 0.05. torch 2.14.0, torchmetrics 1.9.0, current `main`.

| detector | AUPR as shipped | average precision |
|---|---|---|
| perfect, distinct scores | 0.9952 | 1.0000 |
| random, distinct scores | 0.0474 | 0.0433 |
| constant 0.5 | 0.5250 | 0.0500 |
| all-zeros | 0.5250 | 0.0500 |
| inverted, worst possible | 0.5250 | 0.0500 |

A detector that never fires, a detector that returns a single number, and the worst detector that can be constructed all score 0.5250. An honest random detector scores 0.0474. On this metric the three useless ones outrank the merely uninformative one by a factor of eleven, and they land near the middle of the scale, where a reader reasonably takes 0.5 to mean "no better than chance".

The mechanism is that a degenerate score vector has few distinct values, so the curve has few points, and traversing those points in recall order integrates a shape that has nothing to do with detection quality.

## Reproducing

```python
import torch
from torchmetrics.classification import BinaryPrecisionRecallCurve
from torchmetrics.utilities.compute import auc
from torchmetrics.functional.classification import binary_average_precision

class AUPRAsShipped(BinaryPrecisionRecallCurve):
    def compute(self):
        prec, rec, _ = super().compute()
        return auc(rec, prec, reorder=True)

g = torch.Generator().manual_seed(0)
n, k = 2000, 100
target = torch.zeros(n, dtype=torch.long)
target[torch.randperm(n, generator=g)[:k]] = 1

for name, preds in [
    ("constant", torch.full((n,), 0.5)),
    ("all-zeros", torch.zeros(n)),
    ("inverted", 1.0 - target.float()),
    ("random", torch.rand(n, generator=g)),
]:
    m = AUPRAsShipped()
    m.update(preds, target)
    print(name, float(m.compute()), float(binary_average_precision(preds, target)))
```

## Why this one and why now

The same class of defect one file over, `aupro.py` normalising by `fpr[-1]` rather than by the integration limit, was reported in https://github.com/open-edge-platform/anomalib/issues/1651 and https://github.com/open-edge-platform/anomalib/issues/2442 and fixed by https://github.com/open-edge-platform/anomalib/pull/3115. AUPRO also has `tests/unit/metrics/aupro/test_aupro.py` with a reference implementation. There is no `test_aupr.py` anywhere in the repository.

## Suggested fix

Return step-wise average precision, which is the area under the PR curve and is what `torchmetrics.functional.classification.binary_average_precision` already computes. Add `test_aupr.py` asserting that a perfect ranking returns 1.0 and that constant, all-zeros and inverted detectors land at the prevalence floor rather than at 0.5250.

That changes reported numbers on a published metric, so whether it lands as a straight fix or behind a flag is your call. I can open the PR either way.

I have not used the bug template because its Dataset, Model, Configuration YAML and Logs fields do not apply to a metric computation. https://github.com/open-edge-platform/anomalib/issues/3718 is filed the same way.

### PR DESCRIPTION
**Pull request by mateenali66**
## 📝 Description

- 🛠️ Fixes #3781

`_AUPR.compute()` now returns step-wise average precision from the curve `_compute()` already builds, using the formula from https://github.com/open-edge-platform/anomalib/issues/3781#issuecomment-5677799245. The result matches `binary_average_precision` in torchmetrics and `average_precision_score` in scikit-learn.

The old `auc(rec, prec, reorder=True)` sorted the curve by recall and used the trapezoid rule. At 5% prevalence it gave 0.525 to constant, all-zero and inverted detectors, and 0.047 to a random one.

Reported values change. The `AUPR` docstring example goes from 0.8750 to 1.0, and the `_AUPR` example goes from 0.4899 to 0.4610. With `thresholds=100` the old code returned `nan`.

The 8 tests in the new `tests/unit/metrics/test_aupr.py` fail on `main` and pass on this branch. The rest of `tests/unit/metrics` passes, apart from two CUDA tests in `test_pro.py` that also fail on `main` without a GPU.

## ✨ Changes

Select what type of change your PR is:

- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔄 Refactor (non-breaking change which refactors the code base)
- [ ] ⚡ Performance improvements
- [ ] 🎨 Style changes (code style/formatting)
- [x] 🧪 Tests (adding/modifying tests)
- [x] 📚 Documentation update
- [ ] 📦 Build system changes
- [ ] 🚧 CI/CD configuration
- [ ] 🔧 Chore (general maintenance)
- [ ] 🔒 Security update
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [x] 📚 I have made the necessary updates to the documentation (if applicable).
- [x] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).
- [x] 🏷️ My PR title follows conventional commit format.

For more information about code review checklists, see the [Code Review Checklist](https://github.com/open-edge-platform/anomalib/blob/main/docs/source/markdown/guides/developer/contributing.md).

### CODE REVIEW

**Comment by rajeshgangireddy** [comment]
hi @mateenali66 
Thanks for the PR. I will take a look at it soon.

**Review by rajeshgangireddy — APPROVED** [review]
Thanks for the fixes.

**Review by samet-akcay — APPROVED** [review]
